### PR TITLE
PluginBrowser: Tree layout and search bar

### PR DESCRIPTION
**Pull request by CYBERDEViLNL**
https://github.com/LMMS/lmms/issues/4759

### CODE REVIEW

**Comment by JohannesLorenz** [comment]
Works like a charm, both with test-merges on master and lv2.

I'll do a full code review in the next days.

File: include/PluginBrowser.h [hunk 1]
```diff
@@ -39,18 +41,18 @@ class PluginBrowser : public SideBarWidget
 	PluginBrowser( QWidget * _parent );
 	virtual ~PluginBrowser() = default;
 
-private:
-	QWidget * m_view;
-};
-
-
+protected slots:
```
Comment (JohannesLorenz) [inline include/PluginBrowser.h]: Can you make this `private slots`?

File: src/gui/PluginBrowser.cpp [hunk 2]
```diff
@@ -60,23 +61,89 @@ PluginBrowser::PluginBrowser( QWidget * _parent ) :
...
+	// Hide empty roots
+	updateRootVisibilities();
+}
+
+
+void PluginBrowser::updateRootVisibility( int rootIndex )
+{
+	QTreeWidgetItem * root = m_descTree->topLevelItem( rootIndex );
+	if( root->childCount() )
+		root->setHidden( false );
```
Comment (JohannesLorenz) [inline src/gui/PluginBrowser.cpp]: There are multiple uses of `if` without `{}`. Unfortunatelly, the LMMS coding conventions force us to use curly brackets.

File: src/gui/PluginBrowser.cpp [hunk 3]
```diff
@@ -60,23 +61,89 @@ PluginBrowser::PluginBrowser( QWidget * _parent ) :
...
+void PluginBrowser::onFilterChanged( const QString & filter )
 {
-	QVBoxLayout* layout = new QVBoxLayout(this);
+	int rootCount = m_descTree->topLevelItemCount();
+	for (int rootIndex = 0; rootIndex < rootCount; ++rootIndex)
+	{
+		QTreeWidgetItem * root = m_descTree->topLevelItem( rootIndex );
 
+		int itemCount = root->childCount();
+		for( int itemIndex = 0; itemIndex < itemCount; ++itemIndex )
```
Comment (JohannesLorenz) [inline src/gui/PluginBrowser.cpp]: All `if` and `for` must be followed by a space, e.g. `if ` (sorry, coding conventions).

File: src/gui/PluginBrowser.cpp [hunk 4]
```diff
@@ -60,23 +61,89 @@ PluginBrowser::PluginBrowser( QWidget * _parent ) :
...
+		{
+			QTreeWidgetItem * item = root->child( itemIndex );
+			PluginDescWidget * descWidget = static_cast<PluginDescWidget *>(m_descTree->itemWidget( item, 0));
```
Comment (JohannesLorenz) [inline src/gui/PluginBrowser.cpp]: This line is a bit too long (the rule of thumb is ~80 in LMMS).

File: src/gui/PluginBrowser.cpp [hunk 5]
```diff
@@ -60,23 +61,89 @@ PluginBrowser::PluginBrowser( QWidget * _parent ) :
...
+
+	// Hide empty roots
+	updateRootVisibilities();
+}
+
+
+void PluginBrowser::updateRootVisibility( int rootIndex )
+{
+	QTreeWidgetItem * root = m_descTree->topLevelItem( rootIndex );
+	if( root->childCount() )
```
[the remark this answers is not in the document]
Comment (PhysSong) [reply src/gui/PluginBrowser.cpp]: Even more simpler: `root->setVisible(root->childCount());`.

File: src/gui/PluginBrowser.cpp [hunk 6]
```diff
@@ -109,13 +176,14 @@ PluginDescList::PluginDescList(QWidget *parent) :
 
 	for (const PluginKey& key : pluginKeys)
 	{
-		PluginDescWidget* p = new PluginDescWidget( key, this );
-		p->show();
-		layout->addWidget(p);
+		QTreeWidgetItem * item = new QTreeWidgetItem();
+		if( key.desc->name == QString("lv2instrument") )
```
[the remark this answers is not in the document]
Comment (PhysSong) [reply src/gui/PluginBrowser.cpp]: I think [`QStringLiteral` macro](https://doc.qt.io/qt-5/qstring.html#QStringLiteral) may be useful here. It removes the need for extra allocations.

File: include/PluginBrowser.h [hunk 7]
```diff
@@ -27,6 +27,8 @@
 
 #include <QtCore/QTimer>
 #include <QPixmap>
+#include <QTreeWidget>
+#include <QLineEdit>
```
[the remark this answers is not in the document]
Comment (CYBERDEViLNL) [reply include/PluginBrowser.h]: I can’t seem to get it to work with forward decelerations, incomplete types then. I thought you only can use forward decelerations if you are not going to construct?

File: src/gui/PluginBrowser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CYBERDEViLNL) [reply src/gui/PluginBrowser.cpp]: Nice changed to `root->setHidden( !root->childCount() );` I tried `setVisible` but `QtreeWidgetItem` doesn’t have `setVisible`.

File: include/PluginBrowser.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply include/PluginBrowser.h]: If you remove the Qt headers from `PluginBrowser.h`, the errors should only come from `PluginBrowser.cpp`. If you then add the Qt headers back into the cpp file, the errors should vanish. Let me know if it works.

Basically, forward declarations are allowed as long as you only use them as references or pointers and do not dereference/use them, which is the case in your header.

File: src/gui/PluginBrowser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply src/gui/PluginBrowser.cpp]: @PhysSong What's the advantage of `QStringLiteral`? Is it just to be able to use the (better looking) `operator==`, or is there another reason?

[the remark this answers is not in the document]
Comment (PhysSong) [reply src/gui/PluginBrowser.cpp]: @JohannesLorenz If you want to use `strcmp`, you need to convert `QString`'s UTF-16 data to local 8bit characters. Using `QStringLiteral` allows you to construct a UTF-16 string without overheads from conversions and/or allocations.

[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply src/gui/PluginBrowser.cpp]: @PhysSong True, though in this case both sides of the comparison are only available as 8-bit-`const char*`.

File: include/PluginBrowser.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (CYBERDEViLNL) [reply include/PluginBrowser.h]: Thank you, it appears to work :-D

[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply include/PluginBrowser.h]: @CYBERDEViLNL I think you're finished with this point? You can click "Resolve conversation" below when you're done. When all points are resolved, we know you're ready for the merge. :smiley:

**Comment by JohannesLorenz** [comment]
@CYBERDEViLNL It's OK for me. I'd like to merge. Before that, would you mind squashing your five commits into one? It should be as easy as `git rebase -i 4dce466873b3249c0b1a97f6e3a782062623e065` and use "squash" for all but the first one.

**Comment by CYBERDEViLNL** [comment]
@JohannesLorenz good, I think that went well.

**Pull request merged**